toolchain/check/modifiers_allowed.cpp

```diff
@@ -0,0 +1,47 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/modifiers_allowed.h"
+
+namespace Carbon::Check {
+
+static auto ReportNotAllowed(Context& context, Parse::Node modifier_node,
+                             llvm::StringRef decl_name,
+                             Parse::Node context_node) {
+  CARBON_DIAGNOSTIC(ModifierNotAllowedOn, Error, "`{0}` not allowed on {1}.",
+                    llvm::StringRef, llvm::StringRef);
```

Comment:
```suggestion
                    std::string, std::string);
```

Please ensure diagnostic lifetimes can last the lifetime of the program, versus object lifetime. This typically means that `llvm::StringRef` won't work.